packages/core-app-api/src/routing/RouteTracker.tsx

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { useEffect } from 'react';
+import {
+  createRoutesFromChildren,
+  matchRoutes,
+  useLocation,
+} from 'react-router-dom';
+import {
+  BackstagePlugin,
+  useAnalytics,
+  getComponentData,
+  AnalyticsDomain,
+  RoutableAnalyticsDomain,
+} from '@backstage/core-plugin-api';
+
+type RouteObjects = ReturnType<typeof createRoutesFromChildren>;
+
+/**
+ * Returns an extension domain given the current pathname and a RouteObject
+ * that defines all registered routes in react.
+ *
+ * If no exact match is found, path parts are stripped away, one-by-one, until
+ * a parent-level path matches a route.
+ */
+const getExtensionDomain = (
```

Comment:
There must be an easier way to do this that I'm not aware of...?

Reply:
> _It could be something to explore, as in adding this to the routable extensions instead._

Would the idea be that tracking be performed on render of a routable extension?  I guess in that case, the disadvantage would be that any routes added to the router that are not routable extensions wouldn't be captured?

> _This is how it's used by the RouteResolver to look up a common base RouteRef during routing_

I may need to sit with this one awhile to understand fully.  It's almost like what I want is a Backstage equivalent of react router's `createRoutesFromChildren()`, which returns `BackstageRouteObjects[] | null`, and which I can just run a simple `matchRoutes()` on with a given path.  

Am I on the right track, here?

Reply:
> Would the idea be that tracking be performed on render of a routable extension? I guess in that case, the disadvantage would be that any routes added to the router that are not routable extensions wouldn't be captured?

Yep, and yeah it is a disadvantage, it does mean you'll need to stick to the correct type of extension within the route elements if you want tracking.

> Am I on the right track, here?

Yep! We do a big traversal of the elements at startup with a bunch of "collectors". They're responsible for collecting different pieces of information for usage later on. One of them is the [`routeObjectCollector`](https://github.com/backstage/backstage/blob/802f3f106b937794e2da7eeb74086481618bd3d2/packages/core-app-api/src/routing/collectors.tsx#L124), which is essentially `createRoutesFromChildren()`, but the objects also include RouteRefs and could include more stuff like plugins.

Reply:
Think I figured out a reasonable way to do this!  Thanks for the pointers.  See updates